pytest-xdist is not reporting the same nodeid as pytest does

Running `pytest`  and `pytest -n <executors>` does not produce the same nodeid for the tests and therefore this is preventing lastfailed from running properly. Since the nodeids in the cache don't match any test collected, it ends up running tests that are not meant to run.

I have the following workflow:

1. Run the tests that can run in parallel using xdist
2. Run the tests that need to run serially without using xdist's `-n` option
3. Run pytest with the `--last-failed` option to try rerunning the failures to check if they are flaky. This is again without using xdist's `-n` option

Here is the reproducer project structure:

```console
$ ls -1R
.:
pytest.cfg
tests

./tests:
api

./tests/api:
__pycache__
test_bar.py
test_foo.py
```

Here are the file contents:

```console
$ cat tests/api/test_bar.py 
def test_bar_1():
    assert True


def test_bar_2():
    assert True


def test_bar_3():
    assert False

$ cat tests/api/test_foo.py 
import os


def test_foo_fail():
    assert os.environ.get('FOO_PASS', 'false') == 'true'


def test_foo_ok():
    assert True

$ cat pytest.cfg 
[tool:pytest]
python_files = *.py
testpaths = tests/api
```

Considering the project described above, if you run `pytest -n <executors>` it shows the following (mind the tests' nodeids):

```console
$ pytest -v --tb=no -c pytest.cfg -n 2
================================ test session starts ================================
platform linux -- Python 3.7.3, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /home/elyezer/.local/venvs/pytest/bin/python3
cachedir: .pytest_cache
rootdir: /home/elyezer/code/xdist-reproducer, inifile: pytest.cfg, testpaths: tests/api
plugins: xdist-1.29.0, forked-1.0.2
[gw0] linux Python 3.7.3 cwd: /home/elyezer/code/xdist-reproducer
[gw1] linux Python 3.7.3 cwd: /home/elyezer/code/xdist-reproducer
[gw0] Python 3.7.3 (default, May 11 2019, 00:38:04)  -- [GCC 9.1.1 20190503 (Red Hat 9.1.1-1)]
[gw1] Python 3.7.3 (default, May 11 2019, 00:38:04)  -- [GCC 9.1.1 20190503 (Red Hat 9.1.1-1)]
gw0 [5] / gw1 [5]
scheduling tests via LoadScheduling

test_bar.py::test_bar_1 
test_bar.py::test_bar_2 
[gw1] [ 20%] PASSED test_bar.py::test_bar_2 
[gw0] [ 40%] PASSED test_bar.py::test_bar_1 
test_bar.py::test_bar_3 
test_foo.py::test_foo_fail 
[gw0] [ 60%] FAILED test_bar.py::test_bar_3 
test_foo.py::test_foo_ok 
[gw0] [ 80%] PASSED test_foo.py::test_foo_ok 
[gw1] [100%] FAILED test_foo.py::test_foo_fail 

======================== 2 failed, 3 passed in 0.32 seconds =========================
```

But if we run without the `-n` option here is the output:

```console
$ pytest -v --tb=no -c pytest.cfg     
================================ test session starts ================================
platform linux -- Python 3.7.3, pytest-4.6.3, py-1.8.0, pluggy-0.12.0 -- /home/elyezer/.local/venvs/pytest/bin/python3
cachedir: .pytest_cache
rootdir: /home/elyezer/code/xdist-reproducer, inifile: pytest.cfg, testpaths: tests/api
plugins: xdist-1.29.0, forked-1.0.2
collected 5 items                                                                   

tests/api/test_bar.py::test_bar_1 PASSED                                      [ 20%]
tests/api/test_bar.py::test_bar_2 PASSED                                      [ 40%]
tests/api/test_bar.py::test_bar_3 FAILED                                      [ 60%]
tests/api/test_foo.py::test_foo_fail FAILED                                   [ 80%]
tests/api/test_foo.py::test_foo_ok PASSED                                     [100%]

======================== 2 failed, 3 passed in 0.04 seconds =========================
```

As you can see both calls has the same rootdir but produce different nodeids for the tests. And if we check the cache for lastfailed:

```console
$ pytest -c pytest.cfg --cache-show=cache/lastfailed
================================ test session starts ================================
platform linux -- Python 3.7.3, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
rootdir: /home/elyezer/code/xdist-reproducer, inifile: pytest.cfg, testpaths: tests/api
plugins: xdist-1.29.0, forked-1.0.2
cachedir: /home/elyezer/code/xdist-reproducer/.pytest_cache
------------------------ cache values for 'cache/lastfailed' ------------------------
cache/lastfailed contains:
  {'test_bar.py::test_bar_3': True,
   'test_foo.py::test_foo_fail': True,
   'tests/api/test_bar.py::test_bar_3': True,
   'tests/api/test_foo.py::test_foo_fail': True}

=========================== no tests ran in 0.00 seconds ============================
```

`pytest-xdist` test nodeid generation is not consistent with plain pytest nodeid generation and this inconsistency is breaking plugins such as the `lastfailed` that relies on the nodeids to do its job.

Hi @elyezer, 

Really sorry for the delay in answering this!

Thanks a lot for the detailed report, we appreciate it.

The problem is related to the `testpaths` option; commenting that out produces the correct node ids:

```
 λ pytest -v --tb=no -c pytest.cfg -n 2
======================== test session starts ========================
platform win32 -- Python 3.6.3, pytest-4.6.4.dev15+g09dee292c.d20190630, py-1.5.2, pluggy-0.12.0 -- x:\pytest-xdist\.env36\scripts\python.exe
cachedir: .pytest_cache
rootdir: x:\pytest-xdist\.tmp\node-bug, inifile: pytest.cfg
plugins: forked-0.2, xdist-1.28.1.dev9+g407d6e1
[gw0] win32 Python 3.6.3 cwd: x:\pytest-xdist\.tmp\node-bug
[gw1] win32 Python 3.6.3 cwd: x:\pytest-xdist\.tmp\node-bug
[gw0] Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 18:11:49) [MSC v.1900 64 bit (AMD64)]
[gw1] Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 18:11:49) [MSC v.1900 64 bit (AMD64)]
gw0 [5] / gw1 [5]
scheduling tests via LoadScheduling

tests/api/test_bar.py::test_bar_2
tests/api/test_bar.py::test_bar_1
[gw0] [ 20%] PASSED tests/api/test_bar.py::test_bar_1
[gw1] [ 40%] PASSED tests/api/test_bar.py::test_bar_2
tests/api/test_bar.py::test_bar_3
tests/api/test_foo.py::test_foo_fail
[gw0] [ 60%] FAILED tests/api/test_bar.py::test_bar_3
tests/api/test_foo.py::test_foo_ok
[gw0] [ 80%] PASSED tests/api/test_foo.py::test_foo_ok
[gw1] [100%] FAILED tests/api/test_foo.py::test_foo_fail

================ 2 failed, 3 passed in 1.29 seconds =================
``` 

Definitely a bug, thanks for the report.

Actually the underlying cause is the `-c` flag. Unfortunately we have a technical debt regarding config initialization (tracked in https://github.com/pytest-dev/pytest/projects/2).

For now a workaround is to rename your `pytest.cfg` to `setup.cfg` (the standard name), and then the node ids are generated as expected:

```
λ pytest -v --tb=no -n2
======================== test session starts ========================
platform win32 -- Python 3.6.3, pytest-4.6.4.dev15+g09dee292c.d20190630, py-1.5.2, pluggy-0.12.0 -- x:\pytest-xdist\.env36\scripts\python.exe
cachedir: .pytest_cache
rootdir: x:\pytest-xdist\.tmp\node-bug, inifile: setup.cfg, testpaths: tests/api
plugins: forked-0.2, xdist-1.28.1.dev9+g407d6e1
[gw0] win32 Python 3.6.3 cwd: x:\pytest-xdist\.tmp\node-bug
[gw1] win32 Python 3.6.3 cwd: x:\pytest-xdist\.tmp\node-bug
[gw0] Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 18:11:49) [MSC v.1900 64 bit (AMD64)]
[gw1] Python 3.6.3 (v3.6.3:2c5fed8, Oct  3 2017, 18:11:49) [MSC v.1900 64 bit (AMD64)]
gw0 [5] / gw1 [5]
scheduling tests via LoadScheduling

tests/api/test_bar.py::test_bar_2
[gw1] [ 20%] PASSED tests/api/test_bar.py::test_bar_2
tests/api/test_bar.py::test_bar_1
[gw0] [ 40%] PASSED tests/api/test_bar.py::test_bar_1
tests/api/test_foo.py::test_foo_fail
tests/api/test_bar.py::test_bar_3
[gw0] [ 60%] FAILED tests/api/test_bar.py::test_bar_3
[gw1] [ 80%] FAILED tests/api/test_foo.py::test_foo_fail
tests/api/test_foo.py::test_foo_ok
[gw1] [100%] PASSED tests/api/test_foo.py::test_foo_ok

================ 2 failed, 3 passed in 1.05 seconds =================
```

--- a/changelog/448.feature.rst
+++ b/changelog/448.feature.rst
@@ -1,0 +1,9 @@
+Initialization between workers and master nodes is now more consistent, which fixes a number of
+long-standing issues related to startup with the ``-c`` option.
+
+Issues:
+
+* `#6 <https://github.com/pytest-dev/pytest-xdist/issues/6>`__: Poor interaction between ``-n#`` and ``-c X.cfg``
+* `#445 <https://github.com/pytest-dev/pytest-xdist/issues/445>`__: pytest-xdist is not reporting the same nodeid as pytest does
+
+This however only works with **pytest 5.1 or later**, as it required changes in pytest itself.

--- a/src/xdist/remote.py
+++ b/src/xdist/remote.py
@@ -15,6 +15,8 @@ import _pytest.hookspec
 import pytest
 from execnet.gateway_base import dumps, DumpError
 
+from _pytest.config import _prepareconfig, Config
+
 
 class WorkerInteractor(object):
     def __init__(self, config, channel):
@@ -211,18 +213,18 @@ def getinfodict():
 
 
 def remote_initconfig(option_dict, args):
-    from _pytest.config import Config
-
     option_dict["plugins"].append("no:terminal")
-    config = Config.fromdictargs(option_dict, args)
+    return Config.fromdictargs(option_dict, args)
+
+
+def setup_config(config, basetemp):
     config.option.looponfail = False
     config.option.usepdb = False
     config.option.dist = "no"
     config.option.distload = False
     config.option.numprocesses = None
     config.option.maxprocesses = None
-    config.args = args
-    return config
+    config.option.basetemp = basetemp
 
 
 if __name__ == "__channelexec__":
@@ -239,7 +241,13 @@ if __name__ == "__channelexec__":
     os.environ["PYTEST_XDIST_WORKER"] = workerinput["workerid"]
     os.environ["PYTEST_XDIST_WORKER_COUNT"] = str(workerinput["workercount"])
 
-    config = remote_initconfig(option_dict, args)
+    if hasattr(Config, "InvocationParams"):
+        config = _prepareconfig(args, None)
+    else:
+        config = remote_initconfig(option_dict, args)
+        config.args = args
+
+    setup_config(config, option_dict.get("basetemp"))
     config._parser.prog = os.path.basename(workerinput["mainargv"][0])
     config.workerinput = workerinput
     config.workeroutput = {}

--- a/src/xdist/workermanage.py
+++ b/src/xdist/workermanage.py
@@ -186,6 +186,8 @@ def make_reltoroot(roots, args):
     for arg in args:
         parts = arg.split(splitcode)
         fspath = py.path.local(parts[0])
+        if not fspath.exists():
+            continue
         for root in roots:
             x = fspath.relto(root)
             if x or fspath == root:
@@ -236,10 +238,14 @@ class WorkerController(object):
     def setup(self):
         self.log("setting up worker session")
         spec = self.gateway.spec
-        args = self.config.args
+        if hasattr(self.config, "invocation_params"):
+            args = [str(x) for x in self.config.invocation_params.args or ()]
+            option_dict = {}
+        else:
+            args = self.config.args
+            option_dict = vars(self.config.option)
         if not spec.popen or spec.chdir:
             args = make_reltoroot(self.nodemanager.roots, args)
-        option_dict = vars(self.config.option)
         if spec.popen:
             name = "popen-%s" % self.gateway.id
             if hasattr(self.config, "_tmpdirhandler"):

--- a/testing/acceptance_test.py
+++ b/testing/acceptance_test.py
@@ -559,6 +559,30 @@ def test_fixture_teardown_failure(testdir):
     assert result.ret
 
 
+def test_config_initialization(testdir, pytestconfig):
+    """Ensure workers and master are initialized consistently. Integration test for #445"""
+    if not hasattr(pytestconfig, "invocation_params"):
+        pytest.skip(
+            "requires pytest >=5.1 (config has no attribute 'invocation_params')"
+        )
+    testdir.makepyfile(
+        **{
+            "dir_a/test_foo.py": """
+                def test_1(): pass
+        """
+        }
+    )
+    testdir.makefile(
+        ".ini",
+        myconfig="""
+        [pytest]
+        testpaths=dir_a
+    """,
+    )
+    result = testdir.runpytest("-n2", "-c", "myconfig.ini", "-v")
+    result.stdout.fnmatch_lines(["dir_a/test_foo.py::test_1*"])
+
+
 @pytest.mark.parametrize("when", ["setup", "call", "teardown"])
 def test_crashing_item(testdir, when):
     """Ensure crashing item is correctly reported during all testing stages"""

